core/shacl/src/main/java/org/eclipse/rdf4j/plan/GroupBy.java

```diff
@@ -0,0 +1,98 @@
+/*******************************************************************************
+ * Copyright (c) 2016 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+
+package org.eclipse.rdf4j.plan;
+
+import org.eclipse.rdf4j.common.iteration.CloseableIteration;
+import org.eclipse.rdf4j.model.Value;
+import org.eclipse.rdf4j.sail.SailException;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * @author Heshan Jayasinghe
+ */
+public class GroupBy implements GroupPlanNode {
+
+	PlanNode leftjoinnode;
+
+	HashMap<Value, List<List<Value>>> hashMap = new LinkedHashMap<>();
+
+	int skipTupleItems;
+
+	public GroupBy(PlanNode outerLeftJoin, int skipTupleItems) {
+		leftjoinnode = outerLeftJoin;
+		this.skipTupleItems = skipTupleItems;
+	}
+
+	@Override
+	public CloseableIteration<List<Tuple>, SailException> iterator() {
+		CloseableIteration<Tuple, SailException> leftJoinIterator = leftjoinnode.iterator();
+		while (leftJoinIterator.hasNext()) {
+			Tuple leftjointuple = leftJoinIterator.next();
+			boolean status = true;
+			List<List<Value>> values1 = hashMap.computeIfAbsent(leftjointuple.line.get(0),
+					k -> new ArrayList<List<Value>>());
+			List<Value> sublist = leftjointuple.line.subList(skipTupleItems - 1,
+					leftjointuple.line.size() - 1);
+			if (sublist.size() > 0) {
+				values1.add(sublist);
+			}
+		}
+
+		return new CloseableIteration<List<Tuple>, SailException>() {
+
+			Iterator<Map.Entry<Value, List<List<Value>>>> hashmapiterator = hashMap.entrySet().iterator();
+
+			@Override
+			public void close()
+					throws SailException
+			{
+
+			}
+
+			@Override
+			public boolean hasNext()
+					throws SailException
+			{
+				return hashmapiterator.hasNext();
+			}
+
+			@Override
+			public List<Tuple> next()
+					throws SailException
+			{
+				return hashmapiterator.next().getValue().stream().map(Tuple::new).collect(
+						Collectors.toList());
```

Comment:
This materializes the entire grouping in main memory, which will likely not scale well (though I'm not claiming the existing SPARQL impl is any better in that respect). I'm also slightly at a loss as to why you need a separate implementation of this (and several other operators): why doesn't any of this reuse the existing algebra classes?

Reply:
There is no reuse at the moment no. Biggest factor is that it's still unclear what is needed by the SHACL engine. The future of the SHACL engine will require a query engine that can work against three datasources: added triples, removed triples, all triples. I have used Stardog's validation engine quite a lot recently, and it converts OWL to SPARQL and validates the entire database. The number one thing that is impossible to do efficiently is maxCount and minCount with large numbers (>10). These types of validations need to take into account the changed data in the transaction so that a minimal amount of data needs to be fetched from the underlying store for validation.

As for the materialization, this isn't the only place that this is being done. Heshan didn't have time to implement the volcano model properly, so I let him cut some corners for both group by and join so that he would be able to complete support for minCount within the 3-4 months that has been Google Summer of Code.